routes/api.js

```diff
@@ -0,0 +1,166 @@
+// External dependencies
+const path = require('path');
+const express = require('express');
+var router = express.Router();
+
+
+module.exports = function(io, db) {
+  // Routing
+  router.route('/')
+    .get(GETapiGuide);
+
+  router.route('/bobs')
+    .get(GETallBobs)
+    .post(POSTcreateNewBob);
+
+  router.route('/bobs/:bobid')
+    .get(GETbob)
+    .post(function(req, res) {
+      res.status(405)
+        .set('Access-Control-Allow-Methods', 'GET, PUT, DELETE')
+        .send("error: cannot POST, use PUT to edit bobs");
+    })
+    .put(PUTbob)
+    .delete(ensureAuthenticated, DELETEbob);
+
+  router.route('/flavors')
+    .get(GETflavors);
+
+  router.route('/flavors/:flavorname')
+    .get(GETflavor);
+
+  router.route('/tags')
+    .get(GETtags);
+
+  router.route('/tags/:tagid')
+    .get(GETtag);
+
+
+  // API functions
+  function ensureAuthenticated(req, res, next) {
+    if(req.query.auth === 'hunter2'){
+      next();
+    } else {
+      res.status(401).send("User not authenticated");
+    }
+  }
+
+  function GETapiGuide(req, res) {
+    res.sendFile(path.join(__dirname, '..', '/templates/api.html'));
+  }
+
+  function GETallActiveBobs(req, res) {
+    db.Bob.getActiveBobs().then(function(bobs) {
+      res.send(bobs);
+    });
+  }
+
+  function GETallBobs(req, res) {
+    db.Bob.getBobs().then(function(bobs) {
+      res.send(bobs);
+    });
+  }
+
+  function POSTcreateNewBob(req, res, next) {
+    if(!req.body.endDate){
+      req.body.endDate = Date.now() + 2 * 60 * 60 * 12; // Default to two days
+    }
+
+    var bob = {
+      data:      req.body.data,
+      flavor:    req.body.flavor,
+      startDate: req.body.startDate,
+      endDate:   req.body.endDate,
+      tags:      req.body.tags
+    };
+
+    // Save bob in db
+    db.Bob.saveBob(bob).then(function success(bobData) {
+      // Send to all boards
+      io.emit('add_element', bobData);
+      res.send("success");
+    }, function error(err) {
+      res.status(500).send(err);
+    });
+  }
+
+  // Sends back one bob by id
+  function GETbob(req, res) {
+    db.Bob.getOneBob({ _id: db.ObjectId(req.params.bobid) } ).then(function success(data) {
+      res.send(data);
+    }, function error(err) {
+      res.status(500).send(err);
+    });
+  }
+
+  // Updates an existing bob
+  function PUTbob(req, res) {
+    var bob = {
+      _id:			 db.ObjectId(req.body.id),
```

Comment:
_id is used in db.Bob.updateBob, but is not actually written. Same with bob.flavor.

I added comments to explain this